### ISSUE TITLE
[autoscaler][gcp] Worker "Could not open the configuration file" on "[5/7] Initalizing command runner"   

### ISSUE THREAD

**Issue by Mathieu-Prouveur** [ray-project/ray#16708]
First off, thanks for the amazing library!

### What is the problem?

The full example works fine for me https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/gcp/example-full.yaml , however when changing the public docker image to the same docker image hosted on a private repository, the image can't be pulled by any worker.
HeadNode instanciates just fine, it can pull the image on step  "[5/7] Initalizing command runner" , but the workers can't for some reason.

```
2021-06-28 03:17:35,796	INFO log_timer.py:27 -- NodeUpdater: ray-default-worker-98520614: Initialization commands succeeded [LogTimer=14226ms]
2021-06-28 03:17:35,796	INFO updater.py:394 -- [5/7] Initalizing command runner
2021-06-28 03:17:35,797	VINFO command_runner.py:509 -- Running `command -v docker || echo 'NoExist'`
2021-06-28 03:17:35,797	VVINFO command_runner.py:512 -- Full command is `ssh -tt -i ~/ray_bootstrap_key.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_070dd72385/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@10.138.15.213 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (command -v docker || echo '"'"'NoExist'"'"')'`

==> /tmp/ray/session_latest/logs/monitor.err <==
Shared connection to 10.138.15.213 closed.

==> /tmp/ray/session_latest/logs/monitor.out <==
2021-06-28 03:17:37,285	VINFO command_runner.py:509 -- Running `docker pull us.gcr.io/PROJECT_ID/ray`
2021-06-28 03:17:37,285	VVINFO command_runner.py:512 -- Full command is `ssh -tt -i ~/ray_bootstrap_key.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_070dd72385/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@10.138.15.213 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker pull us.gcr.io/PROJECT_ID/ray)'`
Using default tag: latest

==> /tmp/ray/session_latest/logs/monitor.err <==
Shared connection to 10.138.15.213 closed.

==> /tmp/ray/session_latest/logs/monitor.out <==
WARNING: Could not open the configuration file: [/home/ubuntu/.config/gcloud/configurations/config_default].
Error response from daemon: unauthorized: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication
```

It looks to me as if they can't access the file meant to give them the authentification access to google cloud (/home/ubuntu/.config/gcloud/configurations/config_default). I tried to give the workers access to the config file by adding the following lines :
```yaml
initialization_commands:
  - sleep 1 ; [ -f /home/ubuntu/.config/gcloud/configurations/config_default ] && echo "file found" && sudo chown 777 /home/ubuntu/.config/gcloud/configurations/config_default || echo "file not found"
```
But this did not work as no file /home/ubuntu/.config/gcloud/configurations/config_default can be found during initialization step from the workers ("file not found" is printed).

### Reproduction (REQUIRED)
Here's the Dockerfile for reproduction  : `FROM rayproject/ray:latest-cpu`
Then I do a `docker push us.gcr.io/PROJECT_ID/ray` to host it on GCP.
I run a `ray up cluster.yaml` with the following file, and I check the workers status by running a `ray monitor cluster.yaml` (the error above happens at every worker setup attempt) : 
```yaml
# An unique identifier for the head node and workers of this cluster.
cluster_name: default

# The maximum number of workers nodes to launch in addition to the head
# node.
max_workers: 2
min_workers: 2
# The autoscaler will scale up the cluster faster with higher upscaling speed.
# E.g., if the task requires adding more nodes then autoscaler will gradually
# scale up the cluster in chunks of upscaling_speed*currently_running_nodes.
# This number should be > 0.
upscaling_speed: 2.0

# This executes all commands on all nodes in the docker container,
# and opens all the necessary ports to support the Ray cluster.
# Empty string means disabled.
docker:
  image: "us.gcr.io/PROJECT_ID/ray" 
  # image: rayproject/ray:latest-cpu   # works fine with this image
  container_name: "ray_container"
  # If true, pulls latest version of image. Otherwise, `docker run` will only pull the image
  # if no cached version is present.
  pull_before_run: true
  run_options: []  # Extra options to pass into "docker run"

  # Example of running a GPU head with CPU workers
  # head_image: "rayproject/ray-ml:latest-gpu"
  # Allow Ray to automatically detect GPUs

  # worker_image: "rayproject/ray-ml:latest-cpu"
  # worker_run_options: []

# If a node is idle for this many minutes, it will be removed.
idle_timeout_minutes: 15

# Cloud-provider specific configuration.
provider:
    type: gcp
    region: us-west1
    availability_zone: us-west1-a
    project_id: PROJECT_ID # Globally unique project id

# How Ray will authenticate with newly launched nodes.
auth:
    ssh_user: ubuntu
# By default Ray creates a new private keypair, but you can also use your own.
# If you do so, make sure to also set "KeyName" in the head and worker node
# configurations below. This requires that you have added the key into the
# project wide meta-data.
#    ssh_private_key: /path/to/your/key.pem

# Tell the autoscaler the allowed node types and the resources they provide.
# The key is the name of the node type, which is just for debugging purposes.
# The node config specifies the launch config and physical instance type.
available_node_types:
    ray_head_default:
        # The minimum number of worker nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 0
        # The maximum number of worker nodes of this type to launch.
        # This takes precedence over min_workers.
        max_workers: 0
        # The resources provided by this node type.
        resources: {"CPU": 2}
        # Provider-specific config for the head node, e.g. instance type. By default
        # Ray will auto-configure unspecified fields such as subnets and ssh-keys.
        # For more documentation on available fields, see:
        # https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert
        node_config:
            machineType: n1-standard-2
            disks:
              - boot: true
                autoDelete: true
                type: PERSISTENT
                initializeParams:
                  diskSizeGb: 100
                  # See https://cloud.google.com/compute/docs/images for more images
                  sourceImage: projects/deeplearning-platform-release/global/images/family/common-cpu

            # Additional options can be found in in the compute docs at
            # https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert

            # If the network interface is specified as below in both head and worker
            # nodes, the manual network config is used.  Otherwise an existing subnet is
            # used.  To use a shared subnet, ask the subnet owner to grant permission
            # for 'compute.subnetworks.use' to the ray autoscaler account...
            # networkInterfaces:
            #   - kind: compute#networkInterface
            #     subnetwork: path/to/subnet
            #     aliasIpRanges: []
    ray_worker_small:
        # The minimum number of worker nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 2
        # The maximum number of worker nodes of this type to launch.
        # This takes precedence over min_workers.
        max_workers: 2
        # The resources provided by this node type.
        resources: {"CPU": 2}
        # Provider-specific config for the head node, e.g. instance type. By default
        # Ray will auto-configure unspecified fields such as subnets and ssh-keys.
        # For more documentation on available fields, see:
        # https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert
        node_config:
            machineType: n1-standard-2
            disks:
              - boot: true
                autoDelete: true
                type: PERSISTENT
                initializeParams:
                  diskSizeGb: 50
                  # See https://cloud.google.com/compute/docs/images for more images
                  sourceImage: projects/deeplearning-platform-release/global/images/family/common-cpu
            # Run workers on preemtible instance by default.
            # Comment this out to use on-demand.
            scheduling:
              - preemptible: false

    # Additional options can be found in in the compute docs at
    # https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert

# Specify the node type of the head node (as configured above).
head_node_type: ray_head_default

# Files or directories to copy to the head and worker nodes. The format is a
# dictionary from REMOTE_PATH: LOCAL_PATH, e.g.
file_mounts: {
#    "/path1/on/remote/machine": "/path1/on/local/machine",
#    "/path2/on/remote/machine": "/path2/on/local/machine",
}

# Files or directories to copy from the head node to the worker nodes. The format is a
# list of paths. The same path on the head node will be copied to the worker node.
# This behavior is a subset of the file_mounts behavior. In the vast majority of cases
# you should just use file_mounts. Only use this if you know what you're doing!
cluster_synced_files: []

# Whether changes to directories in file_mounts or cluster_synced_files in the head node
# should sync to the worker node continuously
file_mounts_sync_continuously: False

# Patterns for files to exclude when running rsync up or rsync down
rsync_exclude:
    - "**/.git"
    - "**/.git/**"

# Pattern files to use for filtering out files when running rsync up or rsync down. The file is searched for
# in the source directory and recursively through all subdirectories. For example, if .gitignore is provided
# as a value, the behavior will match git's behavior for finding and using .gitignore files.
rsync_filter:
    - ".gitignore"

# List of commands that will be run before `setup_commands`. If docker is
# enabled, these commands will run outside the container and before docker
# is setup.
initialization_commands: []
#  - sleep 1 ; [ -f /home/ubuntu/.config/gcloud/configurations/config_default ] && echo "file found" && sudo chown 777 /home/ubuntu/.config/gcloud/configurations/config_default || echo "file not found"


# List of shell commands to run to set up nodes.
setup_commands: []
    # Note: if you're developing Ray, you probably want to create a Docker image that
    # has your Ray repo pre-cloned. Then, you can replace the pip installs
    # below with a git checkout <your_sha> (and possibly a recompile).
    # To run the nightly version of ray (as opposed to the latest), either use a rayproject docker image
    # that has the "nightly" (e.g. "rayproject/ray-ml:nightly-gpu") or uncomment the following line:
    # - pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"


# Custom commands that will be run on the head node after common setup.
head_setup_commands:
  - pip install google-api-python-client==1.7.8

# Custom commands that will be run on worker nodes after common setup.
worker_setup_commands: []

# Command to start ray on the head node. You don't need to change this.
head_start_ray_commands:
    - ray stop
    - >-
      ulimit -n 65536;
      ray start
      --head
      --port=6379
      --object-manager-port=8076
      --autoscaling-config=~/ray_bootstrap_config.yaml
# Command to start ray on worker nodes. You don't need to change this.
worker_start_ray_commands:
    - ray stop
    - >-
      ulimit -n 65536;
      ray start
      --address=$RAY_HEAD_IP:6379
      --object-manager-port=8076
head_node: {}
worker_nodes: {}
```
- [x] I have verified that ray autoscaler service account can pull images successfully from the container registry
- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

Any help will be appreciated,

**Comment by richardliaw** [comment]
@ijrsvt can you take a look at this?

**Comment by ijrsvt** [comment]
@Mathieu-Prouveur is your GCR registry protected by some form of authentication?

**Comment by Mathieu-Prouveur** [comment]
Hi, my GCR registry's visibility is set to private, therefore gcloud needs to authenticate to a (service) account with the appropriate permissions. The ray-autoscaler-sa-v1 service account has these permissions so it looks to me as if the head node uses this service account during set up, but that the workers do not authenticate with it and therefore can't pull the image from the private gcr.

**Comment by ijrsvt** [comment]
If you manually SSH into a GCP node, can you directly docker pull? I think
you might need to do a `docker login`. (something like
https://cloud.google.com/container-registry/docs/advanced-authentication#gcloud-helper).
These commands (if needed), would be put in `initialization_commands`.

**Comment by Mathieu-Prouveur** [comment]
I logged to a head node and a worker node using `gcloud beta compute ssh` and I could pull from the head node but not from the worker node. I confirmed using `gcloud auth list` that the head node has the ray-autoscaler service account activated while the workers have no service account activated by default.
I will try to write a key json file using the  `initialization_commands` and log with it as suggested, that feels hacky but I also think that should work, I'll let you know after I've tested it.

**Comment by ijrsvt** [comment]
Ahhh @Mathieu-Prouveur Thanks for this information! It appears that by default the Service Account is **only set up for the head**. Let me try to find a way to set the `Service Account` for Worker Nodes :) 

@DmitriGekhtman For AWS we still only set the `autoscaler` role for the `head_node_type`, correct?

**Comment by DmitriGekhtman** [comment]
For AWS we only set the autoscaler role for the _head node_ and we do this by internally populating the `head_node` field (which is not exposed to the user).

**Comment by ijrsvt** [comment]
Similar to https://github.com/ray-project/ray/issues/13799

**Comment by ijrsvt** [comment]
@Mathieu-Prouveur  https://github.com/ray-project/ray/pull/17027 should allow you to specify 
```
    ray_worker_small:
		# Other Config Options here
        node_config:
			# Other Node Config
	        - email: ray-autoscaler-sa-v1@<project_id>.iam.gserviceaccount.com
    	      scopes:
	          - https://www.googleapis.com/auth/cloud-platform
```

**Comment by DmitriGekhtman** [comment]
Though the above might be giving the workers too many permissions.
One could also manually create a service account with necessary permissions and set it in the worker nodes' config, as above.

**Comment by ijrsvt** [comment]
@Mathieu-Prouveur Just wanted to close the loop here--is this working for you now?

**Comment by AmeerHajAli** [comment]
@ijrsvt , thanks for closing the loop!

**Comment by ijrsvt** [comment]
@Mathieu-Prouveur Please let me know if the fix in #17027 does not fix this.

**Comment by Mathieu-Prouveur** [comment]
It works, thank you! Sorry for the late feedback.
